Use a gtkimcontextsimple to handle composite keys

Konubinix opened the pull request:
In case a character is taken into account by the
gtkimcontextsimple, the key_press event is prevented to avoid
doubling the action on the key press.

This pull request is more a medium for discussion about how to do it properly than something to be merged directly.

They are still some problems to deal with:
- the key_release event is still taken into account even for the dead key.
 - For instance, if I type ^ e, I get the following events:
    - EVENT [29243] KEY_RELEASE '' dead_circumflex
    - EVENT [29243] KEY_PRESS '' 'ê'
    - EVENT [29243] KEY_RELEASE '' 'e'```
 - Instead of the expected:
    - EVENT [29243] KEY_PRESS '' 'ê'
    - EVENT [29243] KEY_RELEASE '' 'ê'
- the modifiers in the event is hardcoded to "". I could not find a use case where it is otherwise. For example, if I press "shift I", I get
  - EVENT [29243] MOD_PRESS 'Shift' Shift
  - EVENT [29243] KEY_PRESS '' 'I'
  - EVENT [29243] KEY_RELEASE '' 'I'
  - EVENT [29243] MOD_RELEASE '' Shift

Even if not clean yet, this patch makes composite keys usable in uzbl.

If I type /^etre to look for the french word être in the current page, without the patch, it looked for etre (without the accent). With the patch, it now looks for être (with the accent).

I can also bind a command to a composite key like
- @cbind  ê   = toggle custom_encoding iso-8859-1 utf-8

mathstuf commented:
Would ignoring `dead_`-prefixed keys for events work?

As for dropping `Shift`, it is special on the GDK side. See:

```
EVENT [24275] KEY_PRESS 'Ctrl' 'k'
EVENT [24275] KEY_RELEASE 'Ctrl' 'k'
```

The reasoning is that `Shift+I` is meaningless since the `Shift` is implicit in the `I`. If it came in as `Shift+I`, that'd work (but cause issues with how we manage modifiers in bindings since the modifiers are considered for the entire bind, not individual keys). Try to bind the Vim-ish `<C-x>y` keystrokes. Fixing `bind.py` to handle this is on my TODO list, but will break…eh, every configuration out there.

keis commented:
filtering dead_* would still leave the problem of getting a key_press for ä but a key_release for a because there is no actually ä key. But this is basically how key events work in X so lets look at what trouble it causes and see if we can solve that.

keis commented on src/gui.c:
```diff
@@ -81,6 +91,27 @@ uzbl_gui_init ()
+    (void)dummy;
+
+    uzbl_events_send (KEY_PRESS, NULL,
+        TYPE_STR, "",
```
What if we copy the state (held modifiers) in every keypress/keyrelease to the user_data of this function. See send_keypress_event for this stuff involved for this.

mathstuf replied:
Just as a note, all `user_data` fields will eventually be (at least) the `UzblCore` object to use.

keis replied:
want to attach the last state to the uzbl object right away then?

sorry. this is going a bit off-topic. I think it's a great idea to always pass around the object. it's just sound software design. But trying to limit user_data to only be the uzblcore is dangerous because everything will need to be a field of uzblcore and some stuff can not easily be a single field like "requested_url" with parallel requests. Some cases might call for small focused structs all containing a pointer to the uzblcore ofc.

mathstuf replied:
It was an informative mention :) . Certainly, some will need to be structs (e.g., any that already use `user_data` such as the `_handler` bits).

Konubinix commented:
About the handling of keys, I hesitating between considering ê a key or another object (a character). Then, I took a look at how emacs deals with keys and found out that it considers everything as a key sequence.

For instance, the documentation of the key sequence ^e is ```ê runs the command self-insert-command, which is an interactive built-in function in `C source code'.```

Handling ê as a normal key was the closest behaviour I could get.

About the handling of key_release, according to the documentation of
GtkIMContext
(https://developer.gnome.org/gtk3/stable/GtkIMContext.html#gtk-im-context-filter-keypress),
the input method should handle both the key press and release. I modified the
code to call gtk_im_context_filter_keypress also for the key release event but
it did not change anything.

Investigating this issue, I found out that the implementation of gtk-im-context-filter-keypress in the xim module is somewhat broken since it explicitly avoid handling keyrelease. See in https://github.com/GNOME/gtk/blob/master/modules/input/gtkimcontextxim.c#L654.

Blaming the file (https://github.com/GNOME/gtk/commit/2589ff07b302c932171d1db86ec29f57a4981788#diff-550cd9a936a8e79fc69cc2efbcb69b00L364) seems to indicate that it is due to the bug #81759 (https://bugzilla.gnome.org/show_bug.cgi?id=81759). Causing key to be duplicated some time.

Now, the key doesn't look duplicated but the key_release event is never triggered... Do you have an idea of how to handle this?

Konubinix commented:
In my humble opinion, it would be nice to keep the same mechanism to tackle modifiers. The fact that Shift and Ctrl are treated differently sounds strange.

In my mind, modifiers should be handled separately from keys since we already handle the mod_press/release events.

Could we just remove the modifier information from the key_press event and guess the modifier when the key is surrounded by a mod_press/mod_release couple?

For example, <Ctrl> k would trigger:
- EVENT [9361] MOD_PRESS Ctrl
- EVENT [9361] KEY_PRESS 'k'
- EVENT [9361] KEY_RELEASE 'k'
- EVENT [9361] MOD_RELEASE Ctrl

<Shift> k would be:
- EVENT [9361] MOD_PRESS Shift
- EVENT [9361] KEY_PRESS 'K'
- EVENT [9361] KEY_RELEASE 'K'
- EVENT [9361] MOD_RELEASE Shift

This would avoid the need to copy a modifier somewhere to get it back in the commit callback.

As for the key_release not handled by the input method, we could work around this by adding another event like CHARACTER_TYPED that would be launched on the commit signal.

We would get

For example, <Ctrl> k would trigger:
- EVENT [9361] MOD_PRESS Ctrl
- EVENT [9361] KEY_PRESS 'k'
- EVENT [9361] CHARACTER_TYPED 'k'
- EVENT [9361] KEY_RELEASE 'k'
- EVENT [9361] MOD_RELEASE Ctrl

<Shift> k would be:
- EVENT [9361] MOD_PRESS Shift
- EVENT [9361] KEY_PRESS 'K'
- EVENT [9361] CHARACTER_TYPED 'k'
- EVENT [9361] KEY_RELEASE 'K'
- EVENT [9361] MOD_RELEASE Shift

The keycmd and binding plugin could bind on character_typed instead of the couple key_press, key_release.

mathstuf commented:
On Tue, Oct 20, 2015 at 12:58:47 -0700, Samuel Loury wrote:
> In my humble opinion, it would be nice to keep the same mechanism to
> tackle modifiers. The fact that Shift and Ctrl are treated differently
> sounds strange.

Blame GDK.

> In my mind, modifiers should be handled separately from keys since we
> already handle the mod_press/release events.
> 
> Could we just remove the modifier information from the key_press event
> and guess the modifier when the key is surrounded by a
> mod_press/mod_release couple?

Then every listener to uzbl needs to know that 'Shift' is already added
to the key value for things like 'K'.

> For example, <Ctrl> k would trigger:
> - EVENT [9361] MOD_PRESS Ctrl
> - EVENT [9361] KEY_PRESS 'k'
> - EVENT [9361] KEY_RELEASE 'k'
> - EVENT [9361] MOD_RELEASE Ctrl
> 
> <Shift> k would be:
> - EVENT [9361] MOD_PRESS Shift
> - EVENT [9361] KEY_PRESS 'K'
> - EVENT [9361] KEY_RELEASE 'K'
> - EVENT [9361] MOD_RELEASE Shift
> 
> This would avoid the need to copy a modifier somewhere to get it back
> in the commit callback.

Now every listener needs to do that state tracking. Since it needs done
anyways, having uzbl-core do it is best.

> The keycmd and binding plugin could bind on character_typed instead of
> the couple key_press, key_release.

I'm fine with a new event, but I think sending the modifier along would
be very helpful to event managers.

keis commented:
I agree the including the modstate with each key press is useful. it's also how X does it so following the model makes sense.

the python EM used to track this but it's in a much better place now that it doesn't have to care about that

Konubinix commented:
@keis, I took into account your comment. Now, the input method triggers a key_press event with the correct modifier. I added a new parameter to differentiate a raw key_press and a key_press resulting from an input method to avoid doubling keys in the keycmd plugin. I updated the keycmd plugin accordingly.

keis commented on src/gui.c:
```diff
@@ -1559,13 +1591,22 @@ send_keypress_event (GdkEventKey *event)
 
     /* Get modifier state including this key press/release. */
     modifiers = get_modifier_mask ((event->type == GDK_KEY_PRESS) ? (state | mod) : (state & ~mod));
+    uzbl.gui_->current_key_modifiers = modifiers;
```
this is opens for a segfault as modifiers is free()d at the end of this function. Saving the int state would be easier and just run get_modifier_mask when needed.

keis commented:
A bit tired so maybe I'm missing something here but what is the reason for sending the filtered key presses? It seems the only thing we do with them in the EM is to ignore them.

Konubinix commented:
@keis, personally, i would not send them. i thought that may be someone would want to use them.

Konubinix commented:
funny, now the shift key stopped working correctly. i cannot enter number anymore (in the azerty keyboard, the numbers need to be entered using the shift key).

i honestly don't have a clue of how to fix this. actually, the insert mode looks like black magic to me. grepping the source, i could only see it mentionned in the config file.

do you undertand why this change could break something in the insert mode (interogation mark)

Konubinix commented:
@keis, i saved the state instead of the modifiers as you suggested. thank you for the comment

keis commented:
Ah sorry. I meant the bit where you copy the modifiers string. the memory pointed to by this will be released at the end of the function.
```uzbl.gui_->current_key_modifiers = modifiers;```

What does setting the event->state to 0 do?

I don't think we should send those unless we have a use for them. Even more so as it introduces a breaking change.

Konubinix commented:
@keis, about the current_key_modifiers, I understood what you meant but I forgot to push the modifications... I will do it in a few hours.

I temporarily set the event->state to 0 for the input method to give me a character, independent from the modifier keys. For instance, if I type "<Ctrl>ê", the input method won't handle it because it is not a printable character (I should double check that). I ask the input method to parse only "ê" and I restore the state after-ward.

I will make the code prevent sending the filtered key press, as you suggested.

Konubinix commented:
Done. Now, the gui does not send the raw key press anymore when the input method deal with them.

The plugin is now reverted to its old state since the is no more need to handle duplicates

mathstuf commented:
+1 Lightly tested here. Please rebase to resolve conflicts.

Konubinix commented:
@mathstuf: Done

keis commented:
untested :+1: . I'll give a spin later.

adding inline comments for a few syntax nitpicks

keis commented on src/gui.c:
```diff
@@ -70,6 +76,12 @@ window_init ();
 plug_init ();
 
+static void
+uzbl_commit_cb(
```
keep on one line if possible

keis commented on src/gui.c:
```diff
@@ -84,6 +96,31 @@ uzbl_gui_init ()
+get_modifier_mask (guint state);
+
+static void
+uzbl_commit_cb (
```
also one line or indented by 4.

Konubinix commented:
@keis. Done

mathstuf commented on src/gui.c:
```diff
@@ -54,6 +56,10 @@ struct _UzblGui {
 #endif
 #endif
 
+    GtkIMContext          *im_context;
```
Going through for style. Usually just one space between the type and the name.

mathstuf commented on src/gui.c:
```diff
@@ -70,6 +76,11 @@ window_init ();
+static void uzbl_commit_cb(
+  GtkIMContext* context,
+  const gchar* str,
+  UzblGui* gui);
```
Current style would be:

```c
static void
uzbl_commit_cb(GtkIMContext *context, const gchar *str, gpointer data);
```

With a cast into the `UzblGui` type in the function. Also would be better named as `uzbl_input_commit_cb`.

mathstuf commented on src/gui.c:
```diff
@@ -1816,9 +1845,15 @@ send_keypress_event (GdkEventKey *event)
+    modifiers = get_modifier_mask (uzbl.gui_->current_key_state);
+
+    // set the state to 0 for the input method to process it correctly. The
+    // modifiers are take into account by the previous code anyway
```
C-style comments please.

mathstuf commented on src/gui.c:
```diff
@@ -1816,9 +1845,15 @@ send_keypress_event (GdkEventKey *event)
+    guint old_state = event->state;
+    event->state = 0;
+    if(gtk_im_context_filter_keypress (uzbl.gui_->im_context, event)) {
+    } else if (event->is_modifier && mod) {
```
Add a comment in here as to what is happening and why the body is blank.

Konubinix commented:
I have done the styling fixes and added a comment about the empty body

mathstuf commented:
Merged with some other minor cleanups I missed before. Thanks!

Konubinix commented:
Thank you. I am now able to remove from my todo list a task that I created on 2012-09-06 :-).

keis commented:
![high five](http://i.giphy.com/10LNj580n9OmiI.gif)
